### ISSUE TITLE
FeatureReqiest/DX: Better custom operation arguments

### ISSUE THREAD

**Issue by Richard87** [api-platform/api-platform#1714]
**Description**  
I have a Entity, and a Input DTO for a custom operation, I want my controller to receive both when it's called.

It works now, but it's a super sketchy hack, and I expect it to break in a bugfix...:

**Example**  
Here is my entity, with my custom operation:
```php
/**
 * @ApiResource(
 *     collectionOperations={
 *          "get",
 *          "post": {"security"="is_granted('ROLE_ADMIN')"},
 *     },
 *     itemOperations={
 *          "get",
 *          "delete":{"access_control"="is_granted('ROLE_ADMIN')"},
 *          "api_partnership_create_seminar"={
 *             "path"="/partnerships/{id}/create_seminar",
 *             "method"="POST",
 *             "controller"=CreateSeminar::class,
 *             "input"=CreateSeminarDTO::class,
 *             "security"="is_granted('ROLE_ADMIN')"
 *         }
 *     }
 * )
 */
class Partnership
{
    ...
}
```

I have this custom action
```php

class CreateSeminar
{
    public function __invoke(Partnership $previous_data, CreateSeminarDTO $data): Partnership
    {
        // Do something with CreateSeminarDTO...
        return $previous_data;
    }

}
```

**Comment by Richard87** [comment]
realizing $previous_data is a clone of the original data, took me compleley by suprise :D 

I would love to see some of the ParamConverter magic from Symfonys router in here, trying to match simple cases

Matching types mostly, since there are no concepts for variable names (yet), or a param converter annotation that could handle more specific cases...

**Comment by mbrodala** [comment]
@Richard87 Can you elaborate on what `CreateSeminarDTO` is exactly? I see here that `$data` and `$previous_data` seem to be basically the same object/data:

https://github.com/api-platform/core/blob/v2.5.9/src/EventListener/ReadListener.php#L119-L120

This gives me an error that both of these must have the same type. So I wonder if this still works for you in version 2.5.9 (currently latest.)

**Comment by Richard87** [comment]
Hi, so my hope was to have the Partnership object referenced by {id} in the url, the CreateSeminarDTO is the action I want to do on the object...

For your reference :) 
```php
class CreateSeminarDTO implements SimpleDTOInterface
{
    public \DateTime $time;
    public string $location;
    /**
     * @var string|int[]
     */
    public array $users;
}
```

**Comment by Richard87** [comment]
Yes, they are the same object initially, but $data gets overwritten later on with the input object, which I could take advantage by later in my controller.

BUT this feels **very** wrong, and prone to break in the future.

**Comment by soyuka** [comment]
Closing as stale per #2987 (no activity since 2021-01-21). This repository is transitioning to a CLI installer. If this still reproduces against a current release of API Platform, please open a new issue on `api-platform/core` with a minimal reproduction.
